### DISCUSSION TITLE
Custom serializer not annotated on variable declaration for DTO-generated class

### DISCUSSION THREAD

**Discussion by runningjustin** [babyfish-ct/jimmer#531] [Q&A] [answered]
I have a DTO with the following annotation:

```
    @Serializable(with = BigDecimalSerializer::class)
    subtotal
```

This references the subtotal field in my Entity, which is defined like this:
```
    @Serializable(with = BigDecimalSerializer::class)
    val subtotal: BigDecimal?
```

When the class gets generated, the variable is declared as private (no annotation), then exposed as public with the annotation.

```
    private var _subtotal: BigDecimal? = null
    ...
    @property:Serializable(
        with = BigDecimalSerializer::class
    )
    public var subtotal: BigDecimal?
        get() = _subtotal
        set(subtotal) {
            _subtotal = subtotal
        }
```
This results in the following compiler error:
`Serializer has not been found for type 'BigDecimal?'. To use context serializer as fallback, explicitly annotate type or property with @Contextual`

Am I annotating in the DTO incorrectly? If not, is this a bug?

**Comment by babyfish-ct** [comment]
Yes, I remove the annotation of private field and move it to public property.

This behavior affects kotlin serialization. is it?

**Reply by runningjustin** [reply]
Yeah - gives that compiler error. I can manually add the annotation to the private var and it seems to work.

![image](https://github.com/babyfish-ct/jimmer/assets/136111044/40e92aaf-a57e-42b5-bb7e-df08773c6556)

**Reply by babyfish-ct** [reply]
I don't know kotlin serialization, can you tell me why this causes compiler error and how does kotlin serialization work?

**Comment by babyfish-ct** [comment]
If DTO language supports explicit annotation target like kotlin

```
Input TransactionCreate {
    @field:Serializable(with = BigDecimalSerializer::class)
    subtotal
}
```

Can it resolve problem? If yes, submit FR please.

**Reply by runningjustin** [reply]
This doesn't seem to work.

![image](https://github.com/babyfish-ct/jimmer/assets/136111044/f67b0e11-7631-4939-aba7-f3500ddb4f30)

Apparently not even valid syntax if added directly to the generated class.
![image](https://github.com/babyfish-ct/jimmer/assets/136111044/0522a576-47b2-4f02-8bef-02bd17176f2f)

**Reply by babyfish-ct** [reply]
Of course, I cannot work because it is a new design which is not implemented yet. I mean if this feature can resolve your problem please add a FR.

However, the `@Serializable` of kotlin serialization does not support field.

**Reply by babyfish-ct** [reply, marked as answer]
OK, now the DTO does not use the implicit backing field of kotlin, it uses explicit backing field whose name start with "_".

If this broken kotlin serializaton and there is no workaround, I will change the DTO to remove explicit backing field.

Howevery, I'm super busy now, please wait a few days.

**Reply by runningjustin** [reply]
Thanks, greatly appreciate it and respect your time. FR submitted: https://github.com/babyfish-ct/jimmer/issues/533

**Comment by runningjustin** [comment]
I pulled down 0.8.126 this morning and I see that the private variables have gone away (good). But now how can I apply my custom serializer? I tried the various syntaxes discussed above but nothing makes its way onto the DTO...

**Comment by babyfish-ct** [comment]
```
    @property:Serializable(
        with = BigDecimalSerializer::class
    )
    public val subtotal: BigDecimal?
```

Does the annotation move to generated DTO?

**Reply by runningjustin** [reply]
No, it doesn't. I tested this with 0.8.126 with the syntax above on my Transaction `@Entity` and it still doesn't affect the generated dto classes.

**Comment by babyfish-ct** [comment]
It cannot be reproduced, I validate it like this

1. Provide annotation which has the same `allowedTargets` for `kotlinx.serialization.Serializable`

![image](https://github.com/babyfish-ct/jimmer/assets/15168900/f81365a1-c5c3-4374-90cb-46c130fbc8db)

2. Use it in DTO language

![image](https://github.com/babyfish-ct/jimmer/assets/15168900/70009060-5b8d-4eae-8f4f-8036840d4931)

3. This is the generated DTO

![image](https://github.com/babyfish-ct/jimmer/assets/15168900/4ca23011-5847-4913-9bf7-dd43c4a8e64e)

**Comment by babyfish-ct** [comment]
Oh, you said entity, not DTO.

**Comment by babyfish-ct** [comment]
Try `0.8.128`

**Reply by runningjustin** [reply]
That did it! Thank you! 🙌
